### PR TITLE
Gran Turismo Improvements

### LINKED ISSUE #10882: Bug: Gran Turismo 5/6 desyncs
Gran Turismo games are very unstable on RPCS3, they all have random softlocks. The only setting that "helps" is using PPU/SPU interpreters which slows the game to a crawl to get a few extra minutes of gameplay.

There are two types of desyncs in these games:
- RSX desyncs: When the screen freezes but the game is still running in the background. Happens a lot in CPUs without TSX instructions, Accurate RSX Reservations helps if your CPU is good, but if it's weak and lacks TSX, you'll have them all the time.
- SPU desyncs: When the graphics still render but the game physics are frozen, so you get to see effects like smoke, rain, flashing lights and stuff like that still working, but the game and physics are frozen in time. Input doesn't work and sound is cut out. It's not specifically an SPU issue, as multiple game threads stop working, both on PPU and SPU.

SPU Desync Example
desync

I've added logs for desyncs too, but they don't print any errors to the log file.

- Gran Turismo 5 Prologue
This game is the most stable, you can play for quite a long time before getting a desync. It's SPU desyncs behave differently compared to the other two games, instead of freezing the physics and input, it only cuts out the audio. If it happens mid-race, you can finish it, but the game will then freeze at the next loading screen and won't save your progress. 
I didn't manage to make the game desync in time to get a log that isn't 2gb (it spams the log a lot), but here it is anyway [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/7184257/RPCS3.log.gz)

- Gran Turismo 5
The first game to get SPU desyncs that actually immediately freeze gameplay, GT5 doesn't care much about how good your CPU is, desyncs may happen quickly or take hours. More cars in a race increases the chance of a quick desync, but if you're really dedicated you can finish the game. Here's the log file: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/7184529/RPCS3.log.gz) 

- Gran Turismo 6
GT6 is the most unstable of them all, along with all other issues (especially in 1.22), desyncs run rampant, being difficult to finish a race without desyncing. It's SPU desyncs behave the same as GT5. Log file: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/7184661/RPCS3.log.gz) 

These tests were done with a Ryzen 5 5600x, GTX 1050 Ti running on Windows 10 21H1.

### PR DESCRIPTION
**Pull request by elad335**
I've been laying low lately because I've been working on imprvements to Gran Turismo 5' stability. After many hours debugging and head-racking here are all the fixes to make GT5 not freeze anymore, enjoy.

One thing to not is that you need RSX FIFO Accuracy set to Atomic and enable Accurate RSX reservation access.
Fixes #10882

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/sys_timer.cpp [hunk 1]
```diff
@@ -267,7 +267,8 @@ error_code _sys_timer_start(ppu_thread& ppu, u32 timer_id, u64 base_time, u64 pe
 	{
 		std::lock_guard lock(timer.mutex);
 
-		if (!lv2_obj::check(timer.port))
+		// LV2: Simple nullptr check, nothing complex (do not use lv2_obj::check here)
```
Comment (kd-11) [inline rpcs3/Emu/Cell/lv2/sys_timer.cpp:270]: This comment can be improved by explaining why lv2_obj::check is incorrect here or if maybe it's just unnecessary.

File: Utilities/Thread.cpp [hunk 2]
```diff
@@ -1509,7 +1510,7 @@ bool handle_access_violation(u32 addr, bool is_writing, ucontext_t* context) noe
 			}
 		}
 
-		if (pf_port_id)
+		if (auto pf_port = idm::get<lv2_obj, lv2_event_port>(pf_port_id); pf_port && pf_port->queue)
```
[the remark this answers is not in the document]
Comment (kd-11) [reply Utilities/Thread.cpp:1513]: Ahh, nvm, that's only relevant for RSX handler which happens before this check.

**Review by kd-11 — COMMENTED** [review]
lgtm. Will wait for neko's review.

File: rpcs3/Emu/Cell/PPUThread.cpp [hunk 3]
```diff
@@ -1732,7 +1733,13 @@ bool ppu_thread::savable() const
 
 void ppu_thread::serialize_common(utils::serial& ar)
 {
-	ar(gpr, fpr, cr, fpscr.bits, lr, ctr, vrsave, cia, xer, sat, nj, prio, optional_savestate_state, vr);
+	ar(gpr, fpr, cr, fpscr.bits, lr, ctr, vrsave, cia, xer, sat, nj);
+
+	// if (ar.is_writing())
+	// {
+	// 	s64 prio prio.load().prio;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/PPUThread.cpp:1740]: what's this?

File: rpcs3/Emu/Cell/lv2/sys_cond.cpp [hunk 4]
```diff
@@ -143,36 +143,60 @@ error_code sys_cond_signal(ppu_thread& ppu, u32 cond_id)
...
+		}
 
-			if (const auto cpu = cond.schedule<ppu_thread>(cond.sq, cond.mutex->protocol))
+		bool finished = true;
+
+		ppu.state += cpu_flag::wait;
+
+		const auto cond = idm::check<lv2_obj, lv2_cond>(cond_id, [&, notify = lv2_obj::notify_all_t()](lv2_cond& cond)
+		{
+			//if (atomic_storage<ppu_thread*>::load(cond.sq))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_cond.cpp:160]: what's this ?

File: rpcs3/Emu/Cell/lv2/sys_cond.cpp [hunk 5]
```diff
@@ -181,46 +205,70 @@ error_code sys_cond_signal_all(ppu_thread& ppu, u32 cond_id)
...
+		}
 
-			for (auto cpu = +cond.sq; cpu; cpu = cpu->next_cpu)
+		bool finished = true;
+
+		ppu.state += cpu_flag::wait;
+
+		const auto cond = idm::check<lv2_obj, lv2_cond>(cond_id, [&, notify = lv2_obj::notify_all_t()](lv2_cond& cond)
+		{
+			//if (atomic_storage<ppu_thread*>::load(cond.sq))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_cond.cpp:222]: and this

File: rpcs3/Emu/Cell/lv2/sys_cond.cpp [hunk 6]
```diff
@@ -229,53 +277,77 @@ error_code sys_cond_signal_to(ppu_thread& ppu, u32 cond_id, u32 thread_id)
...
+		const auto cond = idm::check<lv2_obj, lv2_cond>(cond_id, [&, notify = lv2_obj::notify_all_t()](lv2_cond& cond)
 		{
-			std::lock_guard lock(cond.mutex->mutex);
+			if (!idm::check_unlocked<named_thread<ppu_thread>>(thread_id))
+			{
+				return -1;
+			}
 
-			for (auto cpu = +cond.sq; cpu; cpu = cpu->next_cpu)
+			//if (atomic_storage<ppu_thread*>::load(cond.sq))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_cond.cpp:299]: and this

File: rpcs3/Emu/Cell/lv2/sys_sync.h [hunk 7]
```diff
@@ -106,15 +106,20 @@ struct lv2_obj
...
 	static T* unqueue(T*& first, T* object, T* T::* mem_ptr = &T::next_cpu)
 	{
 		auto it = +first;
 
 		if (it == object)
 		{
 			atomic_storage<T*>::release(first, it->*mem_ptr);
-			atomic_storage<T*>::release(it->*mem_ptr, nullptr);
+
+			if (ModifyNode)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_sync.h:118]:
```suggestion
			if constexpr (ModifyNode)
```

File: rpcs3/Emu/Cell/lv2/sys_sync.h [hunk 8]
```diff
@@ -125,7 +130,12 @@ struct lv2_obj
 			if (next == object)
 			{
 				atomic_storage<T*>::release(it->*mem_ptr, next->*mem_ptr);
-				atomic_storage<T*>::release(next->*mem_ptr, nullptr);
+
+				if (ModifyNode)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_sync.h:134]:
```suggestion
				if constexpr (ModifyNode)
```

File: rpcs3/Emu/Cell/lv2/sys_sync.h [hunk 9]
```diff
@@ -478,5 +500,6 @@ struct lv2_obj
 	// If a notify_all_t object exists locally, postpone notifications to the destructor of it (not recursive, notifies on the first destructor for safety)
 	static thread_local bool g_postpone_notify_barrier;
 
+
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_sync.h:503]: empty line

File: rpcs3/Emu/Cell/lv2/sys_lwcond.cpp [hunk 10]
```diff
@@ -102,125 +155,149 @@ error_code _sys_lwcond_signal(ppu_thread& ppu, u32 lwcond_id, u32 lwmutex_id, u6
...
-					return 0;
+					return -1;
 				}
 			}
 
-			auto result = cpu ? cond.unqueue(cond.sq, cpu) :
-				cond.schedule<ppu_thread>(cond.sq, cond.protocol);
-
-			if (result)
+			//if (atomic_storage<ppu_thread*>::load(cond.sq))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_lwcond.cpp:196]: same here

File: rpcs3/Emu/Cell/lv2/sys_lwcond.cpp [hunk 11]
```diff
@@ -102,125 +155,149 @@ error_code _sys_lwcond_signal(ppu_thread& ppu, u32 lwcond_id, u32 lwmutex_id, u6
...
-						// Respect ordering of the sleep queue
-						mutex->try_own(result, true);
-						auto result2 = mutex->reown<ppu_thread>();
+					if (mode == 2)
+					{
+						static_cast<ppu_thread*>(result)->gpr[3] = CELL_EBUSY;
+					}
 
-						if (result2->state & cpu_flag::again)
+					if (mode != 2)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_lwcond.cpp:231]: can mode change?
otherwise:
```suggestion
					else
```

File: rpcs3/Emu/Cell/lv2/sys_lwcond.cpp [hunk 12]
```diff
@@ -237,80 +314,104 @@ error_code _sys_lwcond_signal_all(ppu_thread& ppu, u32 lwcond_id, u32 lwmutex_id
...
+
+				while (const auto cpu = cond.schedule<ppu_thread>(sq, cond.protocol))
 				{
-					mutex->try_own(cpu, true);
+					if (mode == 2)
+					{
+						static_cast<ppu_thread*>(cpu)->gpr[3] = CELL_EBUSY;
+					}
+
+					if (mode == 1)
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_lwcond.cpp:374]:
```suggestion
					else if (mode == 1)
```

File: rpcs3/Emu/Cell/PPUThread.cpp [hunk 13]
```diff
@@ -3754,7 +3761,7 @@ bool ppu_initialize(const ppu_module& info, bool check_only)
 				settings += ppu_settings::accurate_nj_mode, settings -= ppu_settings::fixup_nj_denormals, fmt::throw_exception("NJ Not implemented");
 
 			// Write version, hash, CPU, settings
-			fmt::append(obj_name, "v5-kusa-%s-%s-%s.obj", fmt::base57(output, 16), fmt::base57(settings), jit_compiler::cpu(g_cfg.core.llvm_cpu));
+			fmt::append(obj_name, "v5-kusa-gt-%s-%s-%s.obj", fmt::base57(output, 16), fmt::base57(settings), jit_compiler::cpu(g_cfg.core.llvm_cpu));
```
[the remark this answers is not in the document]
Comment (AniLeo) [reply rpcs3/Emu/Cell/PPUThread.cpp:3764]: Should be just for the test, before merge this should become just v6 I'm guessing

**Comment by ArbestRiAgain** [comment]
@elad335 Will this also affect GT5 Prologue and GT6?

**Comment by Satan86** [comment]
> @elad335 Will this also affect GT5 Prologue and GT6?

Yes.

**Comment by JuanjoSevilla3** [comment]
> > @elad335¿Afectará esto también a GT5 Prologue y GT6?
> 
> Si.

hello! and it also affects Colin McRae Dirt 2 (Mar 24, 2009 QA Build 197111 BDEMU BUILD)?

File: rpcs3/Emu/Cell/lv2/sys_timer.cpp (hunk 1 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_timer.cpp]: Better explained.

File: rpcs3/Emu/Cell/PPUThread.cpp (hunk 3 above)
Comment (elad335) [reply rpcs3/Emu/Cell/PPUThread.cpp]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_cond.cpp (hunk 4 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_cond.cpp]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_cond.cpp (hunk 5 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_cond.cpp]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_cond.cpp (hunk 6 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_cond.cpp]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_sync.h (hunk 7 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_sync.h]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_sync.h (hunk 8 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_sync.h]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_lwcond.cpp (hunk 10 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_lwcond.cpp]: Fixed.

File: rpcs3/Emu/Cell/lv2/sys_lwcond.cpp (hunk 11 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_lwcond.cpp]: Fixed.

**Comment by solarmystic** [comment]
Just did a quick round of tests to compare performance between this PR and master in other non GT games. 

All games tested performed within margin of error to master. 

(Average/1%/0.1% FPS)

**Demons Souls**
PR - 104/86/83 FPS
Master - 106/88/84 FPS

**God of War 3**
PR - 131/101/70 FPS
Master - 131/102/69 FPS

**God of War Ascension**
PR - 66/59/54 FPS
Master - 66/57/51 FPS

**Persona 5**
PR - 81/60/57 FPS
Master - 82/60/55 FPS

**The Last of Us**
PR - 51/47/41 FPS
Master - 52/48/45 FPS 

Screenshots of results:-
[elad gt5 pr vs master.zip](https://github.com/RPCS3/rpcs3/files/11579173/elad.gt5.pr.vs.master.zip)

**Comment by readywer** [comment]
The audio "desync"(audio cut out) still happens sometimes. The easyest way that works for me is to start World Compact Car Race/ Cape ring two times in a row.

**Comment by hoholee12** [comment]
this pr has eliminated GT6 desyncs for good. this is amazing.

**Comment by possiblyRZ** [comment]
besides the 2 settings you listed, are there other ones that you would recommend using with this build? @elad335

**Comment by possiblyRZ** [comment]
> this pr has eliminated GT6 desyncs for good. this is amazing.

which game version and configs are you using? still getting desyncs in GT6 for me... @hoholee12

**Comment by AniLeo** [comment]
Use support platforms if you need support, don't ping further in this PR

https://github.com/RPCS3/rpcs3/wiki/How-to-ask-for-Support

**Pull request merged**